### PR TITLE
Azure Monitor: Azure Resource Graph

### LINKED ISSUE #17202: Resource Graph queries in Azure Monitor datasource
**What would you like to be added**:

Currently azure monitor datasource supports 3 different data types
* Azure monitor
* Application Insights
* Log analytics

additionally, we would like to have **Azure Resource Graph** as a datasource.  There is a REST API available for this. https://docs.microsoft.com/en-us/rest/api/azure-resourcegraph/

Also look at https://azure.microsoft.com/en-gb/features/resource-graph/

**Why is this needed**:

Azure resource graph APIs provide nice way of getting resources and properties of the azure resources. It would be helpful along with application insights and log analytics when using variables.

### PR DESCRIPTION
**Pull request by yesoreyeram**
**What this PR does / why we need it**:

Currently azure monitor datasource supports 3 different data types

* Azure monitor
* Application Insights
* Log analytics

additionally, this pull request will help to add new type of data source from **"Azure Resource Graph"**.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/17202

**Special notes for your reviewer**:

This is my first PR so feel free to suggest corrections.

@briangann For simplicity, It uses the same credentials as azure monitor. No seperate configuration settings added for **Azure Resource Graph**.  Do we need seperate configuration?

I see that we are already in the process of consolidating the configurations for different azure providers.

# Screenshots:

Sample output

image

Query Editor

image

Using resource graph query in variables

image

### CODE REVIEW

**Comment by yesoreyeram** [comment]
# Sample output

image

# Query Editor

image

**Comment by yesoreyeram** [comment]
# Using resource graph query in variables

image

image

**Comment by yesoreyeram** [comment]
@briangann For simplicity, It uses the same credentials as azure monitor. No seperate configuration settings added for **Azure Resource Graph**.  Do we need seperate configuration?

I see that we are already in the process of consolidating the configurations for different azure providers.

**Comment by briangann** [comment]
We're unifying authentication methods in another PR. Separating the config (hopefully) should not be necessary.  The current config method is cumbersome.

Looking forward to see what this PR does :)

**Comment by maxilampert** [comment]
@briangann Any update on when this gets merged?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
+import _ from 'lodash';
+import {
+  AzureDataSourceJsonData,
+  AzureSubscription as IAzureSubscription,
+  AzureResourceGraphQuery as IAzureResourceGraphQuery,
+} from '../types';
+import { DataSourceInstanceSettings } from '@grafana/ui';
+import { BackendSrv } from 'app/core/services/backend_srv';
+import { TemplateSrv } from 'app/features/templating/template_srv';
+import { IQService } from 'angular';
```
Comment (shavonn) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.ts]: We've removed the IQService from these so this should come out in the constructor and `this.$q` replace with `Promise` where ex:`return Promise.all(promises).then(results => {`

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts [hunk 2]
```diff
@@ -10,25 +11,36 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
...
     super(instanceSettings);
     this.azureMonitorDatasource = new AzureMonitorDatasource(instanceSettings, this.templateSrv);
     this.appInsightsDatasource = new AppInsightsDatasource(instanceSettings, this.templateSrv);
-
     this.azureLogAnalyticsDatasource = new AzureLogAnalyticsDatasource(instanceSettings, this.templateSrv);
+    this.azureResourceGraphDatasource = new AzureResourceGraphDatasource(
+      instanceSettings,
+      this.backendSrv,
+      this.templateSrv,
+      this.$q
```
Comment (shavonn) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts]: Remove as arg.

**Review by shavonn — CHANGES_REQUESTED** [review]
Just these couple of changes and fix tests.

**Comment by yesoreyeram** [comment]
Sure @shavonn will update this next week.

**Comment by yesoreyeram** [comment]
Git rebase messed up my PR with many changes happened in the master meanwhile. I am working on cleanup and making changes to align with master.

**Comment by bergquist** [comment]
Moving this to 7.0 since this PR is not ready for review.

**Pull request closed without merging**